tests/helpers.py

```diff
@@ -21,3 +28,95 @@ def name(self):
     def filename(self):
         # weird flask property
         return self._filename
+
+
+class MalleableAny:
+    def __init__(self, condition):
+        self._condition = condition
+
+    def __eq__(self, other):
+        return self._condition(other)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._condition})"
+
+    def __hash__(self):
+        return None
+
+
+class AnySupersetOf(MalleableAny):
+    def __init__(self, subset_dict):
+        self._subset_dict = subset_dict
+        super().__init__(lambda other: self._subset_dict == {k: v for k, v in other.items() if k in self._subset_dict})
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._subset_dict})"
+
+
+class AnyStringMatching(MalleableAny):
+    _cached_re_compile = staticmethod(lru_cache(maxsize=32)(re.compile))
+
+    def __init__(self, *args, **kwargs):
+        self._regex = self._cached_re_compile(*args, **kwargs)
+        super().__init__(lambda other: isinstance(other, six.string_types) and self._regex.match(other))
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}({self._regex})"
+
+
+@contextmanager
+def assert_log_entry(modules, message, count=1):
+    """A context manager that can be used to wrap a block of code under test to assert that a specific message is
+    logged. Works well with the AnyStringMatching malleable above to assert a message meeting a given format.
+
+    This is designed to most easily assert a single kind of message being logged one or more times, but if you
+    assign the context manager to a variable you can access the `records` attribute to see all log records and make
+    further inspections.
+
+    Example:
+        with AssertLogEntry(message=AnyStringMatching('^My unknown string with a random word \w+$', count=2)) as logs:
+            do_something_that_produces_a_log()
+
+        assert any('some other log message' in record.msg for record in logs.records)
+    """
+    log_catcher = logcapture.LogCapture(names=modules, install=False)
+
+    log_catcher.install()
+
+    try:
+        yield log_catcher
+
+    finally:
+        log_catcher.uninstall_all()
+
+        matching_records = [True for record in log_catcher.records if message == record.msg]
+        assert len(matching_records) == count, f'{len(matching_records)} log records were seen that matched ' \
+                                               f'`{message}`: expected {count}'
+
+
+def assert_external_service_log_entry(service='\w+', description='.+', successful_call=True, extra_modules=None,
```

Comment:
Mayyyyybe call the `service` arg `service_pattern` and `description` `description_pattern`? a bit clearer over what these are intended to be used for...

Reply:
Yeahhhhh I guess I coullldddd. But then the `with` context blocks will probably end up taking two lines, and I don't know if I like that more than the short names and updated docstring.

Reply:
"Unicode-replacement-character unicode-replacement-character" to you too

Reply:
How do these regex's get invoked, this is a thing I haven't seen before.

Reply:
Oh that's a point - the default value here `service='\w+'` is missing the `r` prefix. The `\w` will be being turned by python into a ... something ... character.

Reply:
In which case *how does it work at all* :-D

Reply:
It seems `\w` isn't an escaped sequence in python, but I had to try it to find out.

Reply:
@galund the regexes are applied by the `AnyStringMatching` class that @risicle added to test-utils.